### PR TITLE
Adding fnmatch type regex to SFTPSensor

### LINKED ISSUE #15332: SftpSensor w/ possibility to use RegEx or fnmatch
**Description**

SmartSftpSensor with possibility to search for patterns (RegEx or UNIX fnmatch) in filenames or folders

**Use case / motivation**

I would like to have the possibility to use wildcards and/or regular expressions to look for certain files when using an SftpSensor.  
At the moment I tried to do something like this:

```python
from airflow.providers.sftp.sensors.sftp import SFTPSensor
from airflow.plugins_manager import AirflowPlugin
from airflow.utils.decorators import apply_defaults
from typing import Any

import os
import fnmatch

class SmartSftpSensor(SFTPSensor):
    poke_context_fields = ('path', 'filepattern', 'sftp_conn_id', ) # <- Required fields
    template_fields = ['filepattern', 'path']

    @apply_defaults
    def __init__(
            self, 
            filepattern="",
            **kwargs: Any):

        super().__init__(**kwargs)
        self.filepath = self.path
        self.filepattern = filepattern

    def poke(self, context):
        full_path = self.filepath

        directory = os.listdir(full_path)

        for file in directory:
            if not fnmatch.fnmatch(file, self.filepattern):
                pass
            else:
                context['task_instance'].xcom_push(key='file_name', value=file)
                return True
        return False

    def is_smart_sensor_compatible(self): # <- Required
        result = (
            not self.soft_fail
            and super().is_smart_sensor_compatible()
        )
        return result

class MyPlugin(AirflowPlugin):
    name = "my_plugin"
    operators = [SmartSftpSensor]
```
And I call it by doing

```python

sense_file = SmartSftpSensor(
    task_id='sense_file',
    sftp_conn_id='my_sftp_connection',
    path=templ_remote_filepath,
    filepattern=filename,
    timeout=3
)
```
where path is the folder containing the files and filepattern is a rendered filename with wildcards: `filename = """{{ execution_date.strftime("%y%m%d_%H00??_P??_???") }}.LV1"""`, which is rendered to e.g. `210412_1600??_P??_???.LV1`

but I am still not getting the expected result, as it's not capturing anything.

**Are you willing to submit a PR?**
Yes!

**Related Issues**

I didn't find any

### PR DESCRIPTION
**Pull request by alexkruc**
Adding the ability of regex match ([fnmatch](https://docs.python.org/3/library/fnmatch.html) style regex) to `SFTPSensor`. 
This change to the sensor will make it wait for the first available file, based on the given pattern, and then continue working on this file as usual.

My assumption is that `SFTPSensor` should wait for a single file to be delivered - based on a strict format, or an unknown pattern.

Hope that this PR is good enough! <3 

closes: [#15332](https://github.com/apache/airflow/issues/15332)
related: [#15332](https://github.com/apache/airflow/issues/15332)

### CODE REVIEW

File: airflow/providers/sftp/sensors/sftp.py [hunk 1]
```diff
@@ -34,6 +34,7 @@ class SFTPSensor(BaseSensorOperator):
     Waits for a file or directory to be present on SFTP.
 
     :param path: Remote file or directory path
+    :param file_pattern: The regex pattern that will be used to match the file (fnmatch format)
```
Comment (potiuk) [inline airflow/providers/sftp/sensors/sftp.py]:
```suggestion
    :param file_pattern: The pattern that will be used to match the file (fnmatch format)
```

**Review by potiuk — APPROVED** [review]

**Comment by ksuchy** [comment]
Any update on when this will be released?

**Comment by potiuk** [comment]
Did you check in the sftp provider's changelog, if it has not been released already ?

**Comment by ajinkya-xyz** [comment]
@alexkruc  Thanks for adding this feature was waiting for it for quite some time. 

Found a small bug in it, and have raised[ pull request # 25705 ](https://github.com/apache/airflow/pull/25705).  Let me know if it makes sense or am I making some wrong assumptions.

**Pull request merged**